Approximation Framework Blogpost (#3867)
opensearch-project/project-website: _posts/2025-07-14-OpenSearch-Approximation-Framework.md

---
layout: post
title: "OpenSearch Approximation Framework"
category: blog
tags: [search, performance, opensearch-3-1, opensearch-3-0]
authors:

>     - prudhvigodithi

Reviewer: Can we also add @harshavamsi and Sawan as co-authors as well

date: 2025-07-14
categories:
  - technical-posts
meta_keywords: search, query performance, OpenSearch 3.1, OpenSearch 3.0, OpenSearch performance, query optimization
meta_description: The Approximation Framework is a performance optimization in OpenSearch that introduces early termination during BKD tree traversal. It enhances the efficiency of numeric range, sort, and match_all queries by collecting only the required number of documents. This framework delivers exact results with significantly reduced query latency. By intelligently skipping unnecessary work, it improves search responsiveness, especially in high-throughput or time-series workloads.
---

## Introduction

> 

Reviewer: Lets begin with the user story / use case:

"OpenSearch users often run queries on OpenSearch dashboards or applications to retrieve the most recent logs, metrics, or events - for example, fetching the latest 100 error logs from the past 24 hours, or identifying system metrics over the last 7 days for anomaly detection. These queries often include range filters on time or numeric fields (e.g., @timestamp > now-1d), sorting (typically descending), or match_all queries to power visualizations and alerts. Because these are used interactively or in streaming dashboards, latency and efficiency are critical to user experience."

Author: Updated in my latest commit.

Queries involving range filters, sorting, or match_all on numeric fields are widely used in OpenSearch, especially to retrieve records within a bounded window such as logs from the past 24 hours, metrics from the last 7 days, or events over the last 30 days. These queries are typically non-scoring and are often used as the first step in visualizing time-series or event-based data. For example, users may sort results by a timestamp field in descending order to fetch the latest events, or in ascending order to analyze the earliest entries within a given timeframe.

> While Lucene introduced an optimization (`IndexOrDocValuesQuery`) that intelligently uses doc values instead of scanning the entire index when running non-scoring range queries, the default behavior still traverses all segments and often scores more documents than necessary. This results in unnecessarily processing a large set of documents, when only a small subset of top results (e.g., first 50 or 100 hits) are actually required by the user.

Author: Updated with this suggestion.

* **ASC sort**: Uses `intersectLeft` to traverse from smallest to largest values. This method is default and used for plain range queries.
* **DESC sort**: Uses `intersectRight` to traverse from largest to smallest values.

### Example: Traversal with Approximation:

#### `intersectLeft` Traversal :

The diagram below illustrates how the Approximation Framework performs a BKD tree traversal for a query with:

* Size = 1100
* Range = 10:00 – 10:30

![intersectLeft Traversal - flow diagram](/assets/media/blog-images/2025-07-14-OpenSearch-Approximation-Framework/intersectLeft-traversal.png)

Since the goal is to collect just 1100 matching documents, the traversal short-circuits once that threshold is met.

##### Traversal path:
```
Root → Left1 → Left2 → L1 → L2 → Right2 → L3 → Done
```

* Nodes like Right1, Left3, Right3 and all their children (L5–L8) are entirely skipped since enough documents were already collected from the left side of the tree.

> * This demonstrates how the framework avoids visiting unnecessary subtrees, reducing query latency while still returning accurate top-N results.

Author: The `N` is number of `size` requested docs and top is sorted order. Please correct me if I'm wrong are we using `k` in this context in other blogs or docs?

![big5 desc_sort_timestamp query](/assets/media/blog-images/2025-07-14-OpenSearch-Approximation-Framework/big5-desc_sort_timestamp-query-without-approximation.png)

![big5 desc_sort_timestamp query](/assets/media/blog-images/2025-07-14-OpenSearch-Approximation-Framework/big5-desc_sort_timestamp-query-with-approximation.png)

### http_logs: desc_sort_timestamp query

Seen improvement from **~280ms** to **~15ms**

![http_logs desc_sort_timestamp query](/assets/media/blog-images/2025-07-14-OpenSearch-Approximation-Framework/http_logs-desc_sort_timestamp-query-without-approximation.png)

![http_logs desc_sort_timestamp query](/assets/media/blog-images/2025-07-14-OpenSearch-Approximation-Framework/http_logs-desc_sort_timestamp-query-with-approximation.png)

### http_logs: asc_sort_timestamp query

Seen improvement from **~15ms** to **~8ms**

![http_logs asc_sort_timestamp query](/assets/media/blog-images/2025-07-14-OpenSearch-Approximation-Framework/http_logs-asc_sort_timestamp-query-without-approximation.png)

![http_logs asc_sort_timestamp query](/assets/media/blog-images/2025-07-14-OpenSearch-Approximation-Framework/http_logs-asc_sort_timestamp-query-with-approximation.png)

> Further improvements, particularly in sort query performance, were observed in the `3.1` release and are shared in detail on the [3.1.0 release issue](https://github.com/opensearch-project/opensearch-build/issues/5487#issuecomment-2989202040).

Author: Good idea, we can have this section under `Additional improvements:`
```
Further improvements, particularly in sort query performance, were observed in the OpenSearch 3.1 release and are shared in detail in the [3.1.0 release issue](https://github.com/opensearch-project/opensearch-build/issues/5487#issuecomment-2989202040).
 
This result comes from a single-segment test for `desc_sort_timestamp`. According to [this comment](https://github.com/opensearch-project/OpenSearch/pull/18439#issuecomment-2961325856), a force merge to a single segment with an optimized custom BKD walk (`intersectRight`) reduced P90 latency dramatically from **2111** ms to **6.1** ms. This improvement is made possible by the Approximation Framework, which enables early termination during BKD traversal, efficiently collecting the most relevant documents with minimal overhead.

As shown in [this comment](https://github.com/opensearch-project/OpenSearch/pull/18439#issuecomment-2942212895), which captures benchmark results from the development phase, several query types showed significant improvements: `http_logs`: `desc_sort_size` showed over **80%** improvement, `http_logs`: `desc_sort_timestamp` improved by more than **80%**, and `asc_sort_timestamp` achieved over **80.55%** improvement in performance.
```

As shown in this [comment](https://github.com/opensearch-project/OpenSearch/pull/18439#issuecomment-2942212895), which captures benchmark results from the development phase, several query types showed significant improvements: `http_logs desc_sort_size` saw over **80%** improvement, `http_logs`   `desc_sort_timestamp` improved by more than **80%**, and `asc_sort_timestamp` achieved over **80.55%** improvement in performance.

## Upcoming Improvements:

Here is high level [META issue](https://github.com/opensearch-project/OpenSearch/issues/18619) for next set of enhancements related to Approximation Framework.

The following are some of the upcoming improvements planned for the 3.2.0 release, including extending the Approximation Framework to support all numeric types (Related [PR](https://github.com/opensearch-project/OpenSearch/pull/18530)). Additional performance gains have also been observed in range and sort queries, particularly when tested on the skewed http_logs dataset.

> ![http_logs performance comparison](/assets/media/blog-images/2025-07-14-OpenSearch-Approximation-Framework/http_logs-performance-comparison.png)

Reviewer 2: This organization is a bit unclear: the second sentence on line 185 seems to be a performance result. And the next sections (lines 189-213) seem to also be performance results. Should they be moved to before "Upcoming improvements"?

Author: Sure I will take a look. Also FYI we are an RFC for an upcoming enhancement related to this framework https://github.com/opensearch-project/OpenSearch/issues/18784. I would like to push a commit mentioning this. Thanks

![http_logs performance comparison](/assets/media/blog-images/2025-07-14-OpenSearch-Approximation-Framework/http_logs-performance-comparison.png)

### http_logs: range_with_asc_sort (2.19.1--3.2.0) 

`range_with_asc_sort` queries have shown improvement from **~300ms** to **~30ms**, as shown in the following graph. 

![http_logs range_with_asc_sort query](/assets/media/blog-images/2025-07-14-OpenSearch-Approximation-Framework/http_logs-range_with_asc_sort-query-with-approximation.png)

### http_logs: range_size (2.19.1--3.2.0) 

`range_size` queries have shown improvement **~48ms** to **~8ms**, as shown in the following graph. 

![http_logs range_size query](/assets/media/blog-images/2025-07-14-OpenSearch-Approximation-Framework/http_logs-range_size-query-with-approximation.png)

### http_logs: range_with_desc_sort (2.19.1--3.2.0) 

`range_with_desc_sort` queries have shown improvement from **~312ms** to **~31ms**, as shown in the following graph. 

![http_logs range_with_desc_sort query](/assets/media/blog-images/2025-07-14-OpenSearch-Approximation-Framework/http_logs-range_with_desc_sort-query-with-approximation.png)

### nyc_taxis: desc_sort_passenger_count (3.1.0--3.2.0) 

> `desc_sort_passenger_count` queries have shown improvement from **~17ms** to **~12ms**, as shown in the following graph. 

Reviewer 3: ```suggestion
`desc_sort_passenger_count` queries have improved from **~17 ms** to **~12 ms**, as shown in the following graph. 
```

![nyc_taxis desc_sort_passenger_count query](/assets/media/blog-images/2025-07-14-OpenSearch-Approximation-Framework/nyc_taxis-desc_sort_passenger_count-query-with-approximation.png)

### Additioanl query shapes and types under exploration 

> We're exploring several promising areas for extending the Approximation Framework to additional query types. Here are the types of queries we can target in future releases. 

Reviewer 3: ```suggestion
We're exploring several promising extensions of the Approximation Framework to additional query types. Here are the types of queries we can target in future releases. 
```

#### Term query

The proof of concept for extending the framework to top-level term queries on numeric fields has shown approximately **25%** decrease in latency on the `http_logs` dataset. See [related benchmark results](https://github.com/opensearch-project/OpenSearch/pull/18679#issuecomment-3071292692) and [related issue](https://github.com/opensearch-project/OpenSearch/issues/18620).

### Boolean query

> The [related issue](https://github.com/opensearch-project/OpenSearch/issues/18692) and [RFC](https://github.com/opensearch-project/OpenSearch/issues/18784) provide details about implementing `ApproximateBooleanQuery`, which can optimize both single and multi-clause boolean queries, as proof of concept.

Reviewer 3: ```suggestion
[This related issue](https://github.com/opensearch-project/OpenSearch/issues/18692) and [RFC](https://github.com/opensearch-project/OpenSearch/issues/18784) provide information about implementing `ApproximateBooleanQuery`, which can optimize both single- and multi-clause `boolean` queries, as a proof of concept.
```

### Numeric search_after queries

> During a proof of concept effort outlined [in this issue](https://github.com/opensearch-project/OpenSearch/issues/18546), we observed significant improvements for numeric queries using the `search_after` parameter, which is designed for efficient deep pagination of large datasets. In tests with `asc_sort_with_after_timestamp`, the P90 latency dropped from **194.828** ms to **8.459** ms and from **188.037** ms to **7.09** ms for `desc_sort_with_after_timestamp`.

Reviewer 3: ```suggestion
During a proof-of-concept effort outlined [in this issue](https://github.com/opensearch-project/OpenSearch/issues/18546), we observed significant improvements in numeric queries using the `search_after` parameter, which is designed for efficient deep pagination of large datasets. In tests with `asc_sort_with_after_timestamp`, the P90 latency dropped from **194.828 ms** to **8.459 ms**, and it dropped from **188.037 ms** to **7.09 ms** for `desc_sort_with_after_timestamp`.
```
